common/connectors-config/src/main/java/com/vmware/connectors/common/web/ConnectorRootController.java

```diff
@@ -66,7 +66,7 @@ private void addMetadata(ResourceSupport resource, HttpRequest request) {
 
     private void addCards(ResourceSupport resource, HttpRequest request) {
         String cards = UriComponentsBuilder.fromHttpRequest(request).path("/cards/requests").build().toUriString();
-        resource.add(new Link(cards, "cards"));
```

Comment:
I think you're going to need to support both links until connectors are migrated:

```
[tenant_config - 87067]: java.lang.IllegalStateException: Expected to find link with rel 'cards' in response {"_links":{"metadata":{"href":"http://172.17.100.100:8064/discovery/metadata.json"},"objects":{"href":"http://172.17.100.100:8064/cards/requests"},"image":{"href":"http://172.17.100.100:8064/images/connector.png"}}}!
[tenant_config - 87067]: 	at org.springframework.hateoas.client.Traverson$TraversalBuilder.getAndFindLinkWithRel(Traverson.java:436) ~[spring-hateoas-0.24.0.RELEASE.jar!/:na]
[tenant_config - 87067]: 	at org.springframework.hateoas.client.Traverson$TraversalBuilder.traverseToExpandedFinalUrl(Traverson.java:412) ~[spring-hateoas-0.24.0.RELEASE.jar!/:na]
[tenant_config - 87067]: 	at org.springframework.hateoas.client.Traverson$TraversalBuilder.traverseToLink(Traverson.java:400) ~[spring-hateoas-0.24.0.RELEASE.jar!/:na]
[tenant_config - 87067]: 	at org.springframework.hateoas.client.Traverson$TraversalBuilder.asLink(Traverson.java:384) ~[spring-hateoas-0.24.0.RELEASE.jar!/:na]
[tenant_config - 87067]: 	at com.vmware.platform.tenantconfig.connector.discovery.ConnectorDiscoveryFactory$2.load(ConnectorDiscoveryFactory.java:73) ~[classes!/:1.0-SNAPSHOT]
[tenant_config - 87067]: 	at com.vmware.platform.tenantconfig.connector.discovery.ConnectorDiscoveryFactory$2.load(ConnectorDiscoveryFactory.java:67) ~[classes!/:1.0-SNAPSHOT]
[tenant_config - 87067]: 	at com.google.common.cache.LocalCache$LoadingValueReference.loadFuture(LocalCache.java:3523) ~[guava-23.6-jre.jar!/:na]
[tenant_config - 87067]: 	at com.google.common.cache.LocalCache$Segment.loadSync(LocalCache.java:2249) ~[guava-23.6-jre.jar!/:na]
[tenant_config - 87067]: 	at com.google.common.cache.LocalCache$Segment.lockedGetOrLoad(LocalCache.java:2132) ~[guava-23.6-jre.jar!/:na]
[tenant_config - 87067]: 	at com.google.common.cache.LocalCache$Segment.get(LocalCache.java:2045) ~[guava-23.6-jre.jar!/:na]
[tenant_config - 87067]: 	at com.google.common.cache.LocalCache.get(LocalCache.java:3962) ~[guava-23.6-jre.jar!/:na]
[tenant_config - 87067]: 	at com.google.common.cache.LocalCache.getOrLoad(LocalCache.java:3966) ~[guava-23.6-jre.jar!/:na]
[tenant_config - 87067]: 	at com.google.common.cache.LocalCache$LocalLoadingCache.get(LocalCache.java:4951) ~[guava-23.6-jre.jar!/:na]
[tenant_config - 87067]: 	at com.google.common.cache.LocalCache$LocalLoadingCache.getUnchecked(LocalCache.java:4957) ~[guava-23.6-jre.jar!/:na]
[tenant_config - 87067]: 	at com.vmware.platform.tenantconfig.connector.discovery.ConnectorDiscoveryFactory.getFromCache(ConnectorDiscoveryFactory.java:57) ~[classes!/:1.0-SNAPSHOT]
[tenant_config - 87067]: 	at com.vmware.platform.tenantconfig.connector.discovery.ConnectorDiscoveryFactory.fromTenantConnector(ConnectorDiscoveryFactory.java:51) ~[classes!/:1.0-SNAPSHOT]
[tenant_config - 87067]: 	at com.vmware.platform.tenantconfig.resources.ConnectorResourceAssembler.toResourceInternal(ConnectorResourceAssembler.java:77) ~[classes!/:1.0-SNAPSHOT]
[tenant_config - 87067]: 	at com.vmware.platform.tenantconfig.resources.ConnectorResourceAssembler.lambda$toResource$0(ConnectorResourceAssembler.java:65) ~[classes!/:1.0-SNAPSHOT]
[tenant_config - 87067]: 	at com.vmware.platform.utils.CircuitBreakers$1.run(CircuitBreakers.java:48) ~[platform-common-1.0-SNAPSHOT.jar!/:1.0-SNAPSHOT]
[tenant_config - 87067]: 	at com.netflix.hystrix.HystrixCommand$2.call(HystrixCommand.java:302) ~[hystrix-core-1.5.12.jar!/:1.5.12]
[tenant_config - 87067]: 	at com.netflix.hystrix.HystrixCommand$2.call(HystrixCommand.java:298) ~[hystrix-core-1.5.12.jar!/:1.5.12]
[tenant_config - 87067]: 	at rx.internal.operators.OnSubscribeDefer.call(OnSubscribeDefer.java:46) [rxjava-1.3.8.jar!/:1.3.8]
[tenant_config - 87067]: 	at rx.internal.operators.OnSubscribeDefer.call(OnSubscribeDefer.java:35) [rxjava-1.3.8.jar!/:1.3.8]
[tenant_config - 87067]: 	at rx.internal.operators.OnSubscribeLift.call(OnSubscribeLift.java:48) [rxjava-1.3.8.jar!/:1.3.8]
[tenant_config - 87067]: 	at rx.internal.operators.OnSubscribeLift.call(OnSubscribeLift.java:30) [rxjava-1.3.8.jar!/:1.3.8]
[tenant_config - 87067]: 	at rx.internal.operators.OnSubscribeLift.call(OnSubscribeLift.java:48) [rxjava-1.3.8.jar!/:1.3.8]
[tenant_config - 87067]: 	at rx.internal.operators.OnSubscribeLift.call(OnSubscribeLift.java:30) [rxjava-1.3.8.jar!/:1.3.8]
[tenant_config - 87067]: 	at rx.internal.operators.OnSubscribeLift.call(OnSubscribeLift.java:48) [rxjava-1.3.8.jar!/:1.3.8]
[tenant_config - 87067]: 	at rx.internal.operators.OnSubscribeLift.call(OnSubscribeLift.java:30) [rxjava-1.3.8.jar!/:1.3.8]
[tenant_config - 87067]: 	at rx.Observable.unsafeSubscribe(Observable.java:10327) [rxjava-1.3.8.jar!/:1.3.8]
[tenant_config - 87067]: 	at rx.internal.operators.OnSubscribeDefer.call(OnSubscribeDefer.java:51) [rxjava-1.3.8.jar!/:1.3.8]
[tenant_config - 87067]: 	at rx.internal.operators.OnSubscribeDefer.call(OnSubscribeDefer.java:35) [rxjava-1.3.8.jar!/:1.3.8]
[tenant_config - 87067]: 	at rx.Observable.unsafeSubscribe(Observable.java:10327) [rxjava-1.3.8.jar!/:1.3.8]
[tenant_config - 87067]: 	at rx.internal.operators.OnSubscribeDoOnEach.call(OnSubscribeDoOnEach.java:41) [rxjava-1.3.8.jar!/:1.3.8]
[tenant_config - 87067]: 	at rx.internal.operators.OnSubscribeDoOnEach.call(OnSubscribeDoOnEach.java:30) [rxjava-1.3.8.jar!/:1.3.8]
[tenant_config - 87067]: 	at rx.internal.operators.OnSubscribeLift.call(OnSubscribeLift.java:48) [rxjava-1.3.8.jar!/:1.3.8]
[tenant_config - 87067]: 	at rx.internal.operators.OnSubscribeLift.call(OnSubscribeLift.java:30) [rxjava-1.3.8.jar!/:1.3.8]
[tenant_config - 87067]: 	at rx.Observable.unsafeSubscribe(Observable.java:10327) [rxjava-1.3.8.jar!/:1.3.8]
[tenant_config - 87067]: 	at rx.internal.operators.OperatorSubscribeOn$SubscribeOnSubscriber.call(OperatorSubscribeOn.java:100) [rxjava-1.3.8.jar!/:1.3.8]
[tenant_config - 87067]: 	at com.netflix.hystrix.strategy.concurrency.HystrixContexSchedulerAction$1.call(HystrixContexSchedulerAction.java:56) [hystrix-core-1.5.12.jar!/:1.5.12]
[tenant_config - 87067]: 	at com.netflix.hystrix.strategy.concurrency.HystrixContexSchedulerAction$1.call(HystrixContexSchedulerAction.java:47) [hystrix-core-1.5.12.jar!/:1.5.12]
[tenant_config - 87067]: 	at com.netflix.hystrix.strategy.concurrency.HystrixContexSchedulerAction.call(HystrixContexSchedulerAction.java:69) [hystrix-core-1.5.12.jar!/:1.5.12]
[tenant_config - 87067]: 	at rx.internal.schedulers.ScheduledAction.run(ScheduledAction.java:55) [rxjava-1.3.8.jar!/:1.3.8]
[tenant_config - 87067]: 	at java.util.concurrent.Executors$RunnableAdapter.call(Executors.java:511) [na:1.8.0_144]
[tenant_config - 87067]: 	at java.util.concurrent.FutureTask.run(FutureTask.java:266) [na:1.8.0_144]
[tenant_config - 87067]: 	at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1149) [na:1.8.0_144]
[tenant_config - 87067]: 	at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:624) [na:1.8.0_144]
[tenant_config - 87067]: 	at java.lang.Thread.run(Thread.java:748) [na:1.8.0_144]
[tenant_config - 87067]: 
```

Reply:
Per @rworsnop  suggestion in the APF-1253 ticket, we have to provide fallback in the hero service.

I have made changes in the ```api-gateway``` to add fallback if ```cards``` link is not available.

https://stash.air-watch.com/projects/UFO/repos/app-platform-server/pull-requests/343/diff#tenant-configuration/src/main/java/com/vmware/platform/tenantconfig/connector/discovery/ConnectorDiscoveryFactory.java

Reply:
Just tested in the APF-1253 branch and it's working.

I'm changing the review-feedback to merge the APF PR first.